### PR TITLE
Dormitory features

### PR DESCRIPTION
**Pull request by mrjackv**
Sibling PRs:
* orchestra: https://github.com/revng/orchestra/pull/138
* revng-ui: https://rev.ng/gitlab/revng-private/revng-ui/-/merge_requests/16

### CODE REVIEW

File: include/revng/PipelineDescriptor/PipelineDescriptor.h [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+doc: Data structure representing the structure of the pipeline
+type: struct
+fields:
+  - name: AnalysesLists
+    doc: AnalysesLists available in the pipeline
+    sequence:
+      type: SortedVector
+      elementType: AnalysesList
+  - name: Globals
+    doc: Globals present
```
Comment (aleclearmind) [inline include/revng/PipelineDescriptor/PipelineDescriptor.h:20]: List of globals

File: include/revng/PipelineDescriptor/Analysis.h [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+/* TUPLE-TREE-YAML
+name: Analysis
+doc: Data structure representing an analysis
+type: struct
+fields:
+  - name: Name
+    doc: Analysis' name
+    type: string
+  - name: Arguments
+    doc: Analysis' arguments
```
Comment (aleclearmind) [inline include/revng/PipelineDescriptor/Analysis.h:16]: 1. "Analysis' argument names"?
2. Don't arguments have a type too?

File: include/revng/PipelineDescriptor/AnalysisArgument.h [hunk 3]
```diff
@@ -0,0 +1,32 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+/* TUPLE-TREE-YAML
+name: AnalysisArgument
```
Comment (aleclearmind) [inline include/revng/PipelineDescriptor/AnalysisArgument.h:8]: This seems to be unused, but there's a fixup later on. Backport it.

Also, these are the accepted kinds, they don't have anything to do with analysis arguments.

The anlaysis arguments are options, but the accepted kinds is a property of the analysis, not of the argument.

File: include/revng/PipelineDescriptor/PipelineDescriptor.h [hunk 4]
```diff
@@ -0,0 +1,52 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "revng/PipelineDescriptor/AnalysisArgument.h"
```
Comment (aleclearmind) [inline include/revng/PipelineDescriptor/PipelineDescriptor.h:7]: 1. Push all of this to `revng/Pipeline/Descriptor/`.
2. "Descriptor" is something you "attach" to something, like a label. We can go for `Description` or `Metadata`.

File: include/revng/Pipes/PipelineManager.h [hunk 5]
```diff
@@ -13,6 +13,7 @@
 
 #include "revng/Pipeline/Context.h"
 #include "revng/Pipeline/Runner.h"
+#include "revng/PipelineDescriptor/PipelineDescriptor.h"
```
Comment (aleclearmind) [inline include/revng/Pipes/PipelineManager.h:16]: This include doesn't seem to be necessary since you're using a `std::string` to store the result.

File: lib/Pipes/DescriptorConverter.cpp [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "revng/PipelineDescriptor/AnalysisArgument.h"
+#include "revng/PipelineDescriptor/Container.h"
+#include "revng/Pipes/DescriptorConverter.h"
+
+static pipelineDescriptor::Analysis
+convert(const pipeline::AnalysisWrapper &Analysis) {
```
Comment (aleclearmind) [inline lib/Pipes/DescriptorConverter.cpp:12]: `s/convert/describe/`?

File: lib/Pipes/PipelineManager.cpp [hunk 7]
```diff
@@ -500,3 +507,37 @@ PipelineManager::produceTargets(const llvm::StringRef StepName,
...
+  if (ExecutionDirectory.empty())
+    return llvm::Error::success();
+
+  if (auto EC = llvm::sys::fs::create_directories(ExecutionDirectory); EC) {
+    return llvm::createStringError(EC,
+                                   "Could not create dir %s",
+                                   ExecutionDirectory.c_str());
+  }
+
+  std::string DescriptorPath = joinPath(ExecutionDirectory, "descriptor.yml");
```
Comment (aleclearmind) [inline lib/Pipes/PipelineManager.cpp:530]: `pipeline-description.yml`?

File: lib/Pipes/PipelineManager.cpp [hunk 8]
```diff
@@ -142,6 +143,9 @@ PipelineManager::create(llvm::ArrayRef<std::string> Pipelines,
     return MaybePipeline.takeError();
 
   Manager.recalculateAllPossibleTargets();
+  if (auto Error = Manager.computeDescriptor(); Error)
```
Comment (aleclearmind) [inline lib/Pipes/PipelineManager.cpp:146]: 1. Add a blank line, computing the descriptor is not related to do with `recalculateAllPossibleTargets`.
2. There's some code duplication going on here, can we have a `PipelineManager::finalizeIntialization` `private` method that does `recalculateAllPossibleTargets` and `computeDescriptor` (and maybe other stuff)?

File: python/revng/pipelineDescriptor/__init__.py [hunk 9]
```diff
@@ -0,0 +1,5 @@
+#
```
Comment (aleclearmind) [inline python/revng/pipelineDescriptor/__init__.py:1]: Python uses `snake_case`: `pipeline_descriptor`

File: python/revng/tupletree.py [hunk 10]
```diff
@@ -326,3 +326,47 @@ def factory():
         return TypedList(base_class)
 
     return factory
+
+
+def enum_value_to_index(enum_value: Enum):
```
Comment (aleclearmind) [inline python/revng/tupletree.py:331]: This change seems to be a good candidate for its own commit.

File: scripts/tuple_tree_generator/tuple_tree_generator/templates/struct_impl.cpp.tpl [hunk 11]
```diff
@@ -6,7 +6,9 @@ The notice below applies to the generated files.
 // This file is distributed under the MIT License. See LICENSE.md for details.
 //
 
-#include "revng/Model/Binary.h"
+/** if root_type **/
+#include "/*= generator.user_include_path =*//*= root_type =*/.h"
+/** endif **/
```
Comment (aleclearmind) [inline scripts/tuple_tree_generator/tuple_tree_generator/templates/struct_impl.cpp.tpl:11]: This deserves its own commit

File: typescript/CMakeLists.txt [hunk 12]
```diff
@@ -17,23 +17,37 @@ add_custom_target(
...
-
-add_custom_target(generate-model.ts.tgz ALL
-                  DEPENDS "${CMAKE_BINARY_DIR}/model.ts.tgz")
-
-install(FILES "${CMAKE_BINARY_DIR}/model.ts.tgz"
-        DESTINATION "${CMAKE_INSTALL_PREFIX}/share/typescript")
-
-install(FILES "${CMAKE_BINARY_DIR}/model.ts.tgz"
-        DESTINATION "${CMAKE_INSTALL_PREFIX}/share/node_cache")
+function(typescript_module)
```
Comment (aleclearmind) [inline typescript/CMakeLists.txt:20]: Nice

File: typescript/CMakeLists.txt [hunk 13]
```diff
@@ -17,23 +17,37 @@ add_custom_target(
...
+
+  install(FILES "${CMAKE_BINARY_DIR}/${TARGET_NAME}.ts.tgz"
+          DESTINATION "${CMAKE_INSTALL_PREFIX}/share/typescript")
+
+  install(FILES "${CMAKE_BINARY_DIR}/${TARGET_NAME}.ts.tgz"
+          DESTINATION "${CMAKE_INSTALL_PREFIX}/share/node_cache")
+endfunction()
+
+typescript_module(TARGET_NAME model)
+typescript_module(TARGET_NAME pipelineDescriptor)
```
Comment (aleclearmind) [inline typescript/CMakeLists.txt:50]: TypeScript packages seem to use `kebab-case`

File: include/revng/Pipeline/ContainerFactorySet.h [hunk 14]
```diff
@@ -34,6 +34,18 @@ class ContainerFactorySet {
...
+    auto Search = Entries.find(Name);
+    revng_assert(Search != Entries.end());
+    return *Search;
+  }
+
+  const value_type &find(llvm::StringRef Name) const {
+    auto Search = Entries.find(Name);
+    revng_assert(Search != Entries.end());
+    return *Search;
+  }
```
Comment (aleclearmind) [inline include/revng/Pipeline/ContainerFactorySet.h:47]: `find` methods return an iterator. This looks more like `at()`.

File: include/revng/PipelineC/Prototypes.h [hunk 15]
```diff
@@ -187,6 +196,11 @@ uint64_t rp_manager_step_name_to_index(const rp_manager *manager,
  */
 rp_step *rp_manager_get_step(const rp_manager *manager, uint64_t index);
 
+/**
+ * \return the step with the provided nale, or NULL if not such step existed.
```
Comment (aleclearmind) [inline include/revng/PipelineC/Prototypes.h:168]: *name

File: include/revng/PipelineDescriptor/Analysis.h [hunk 16]
```diff
@@ -16,7 +16,7 @@ type: struct
     doc: Analysis' arguments
     sequence:
       type: SortedVector
-      elementType: string
+      elementType: AnalysisArgument
```
Comment (aleclearmind) [inline include/revng/PipelineDescriptor/Analysis.h]: Squash in the commit introducing this.

File: lib/Pipes/DescriptorConverter.cpp [hunk 17]
```diff
@@ -19,6 +19,8 @@ convert(const pipeline::AnalysisWrapper &Analysis) {
     Argument.Name() = ArgumentNames[I];
     for (auto &Kind : Analysis->getAcceptedKinds(I))
       Argument.AcceptableKinds().insert(Kind->name().str());
+
+    Result.Arguments().insert(Argument);
```
Comment (aleclearmind) [inline lib/Pipes/DescriptorConverter.cpp:23]: Squash

File: python/CMakeLists.txt [hunk 18]
```diff
@@ -122,15 +122,10 @@ python_module(TARGET_NAME revng-python-cli-commands-daemon_self_test
...
     revng/api/_capi.py
-    revng/api/analysis.py
-    revng/api/container.py
     revng/api/errors.py
     revng/api/exceptions.py
     revng/api/invalidations.py
-    revng/api/kind.py
     revng/api/manager.py
-    revng/api/rank.py
-    revng/api/step.py
```
Comment (aleclearmind) [inline python/CMakeLists.txt:134]: Nice!

File: python/revng/daemon/graphql.py [hunk 19]
```diff
@@ -0,0 +1,198 @@
+#
```
Comment (aleclearmind) [inline python/revng/daemon/graphql.py:1]: Is this just a rename or actual new code?
GitHub shows it as new code.

File: include/revng/PipelineC/Prototypes.h [hunk 20]
```diff
@@ -141,19 +141,6 @@ LENGTH_HINT(rp_manager_create_from_string, 3, 2)
  */
 void rp_manager_destroy(rp_manager *manager);
 
-/**
```
Comment (aleclearmind) [inline include/revng/PipelineC/Prototypes.h:144]: Great

**Review by aleclearmind — CHANGES_REQUESTED** [review]

File: include/revng/PipelineDescriptor/Kind.h [hunk 21]
```diff
@@ -0,0 +1,43 @@
...
+doc: Data structure representing a Kind
+type: struct
+fields:
+  - name: Name
+    doc: Kind's name
+    type: string
+  - name: Rank
+    doc: Kind's rank
+    type: string
+  - name: KindParent
```
[the remark this answers is not in the document]
Comment (mrjackv) [reply include/revng/PipelineDescriptor/Kind.h:18]: Due to a limitation of `TTG` the `Parent` field is already used [refrence](https://github.com/revng/revng/blob/develop/scripts/tuple_tree_generator/tuple_tree_generator/templates/struct.h.tpl#L43) so it needs the `Kind` part beforehand

File: include/revng/PipelineDescriptor/Kind.h [hunk 22]
```diff
@@ -0,0 +1,43 @@
...
+    doc: The Kind's parent (if present)
+    type: string
+    optional: true
+  - name: DefinedLocations
+    doc: List of locations that the Kind provides
+    sequence:
+      type: SortedVector
+      elementType: string
```
[the remark this answers is not in the document]
Comment (mrjackv) [reply include/revng/PipelineDescriptor/Kind.h:26]: e.g.
ASM -> `/function/`, `/basic-block/`, `/instruction/`
C -> `/function/`

File: include/revng/PipelineDescriptor/Step.h [hunk 23]
```diff
@@ -0,0 +1,39 @@
...
+
+/* TUPLE-TREE-YAML
+name: Step
+doc: Data structure representing a Step
+type: struct
+fields:
+  - name: Name
+    doc: Step's name
+    type: string
+  - name: StepParent
```
[the remark this answers is not in the document]
Comment (mrjackv) [reply include/revng/PipelineDescriptor/Step.h:15]: See [here](https://github.com/revng/revng/pull/318#discussion_r1304027327)

File: include/revng/PipelineDescriptor/AnalysesList.h [hunk 24]
```diff
@@ -0,0 +1,32 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+/* TUPLE-TREE-YAML
+name: AnalysesList
```
[the remark this answers is not in the document]
Comment (mrjackv) [reply include/revng/PipelineDescriptor/AnalysesList.h:8]: The `PipelineManager` API uses `AnalysesList` (this was discussed when the term was introduced). Since it's a list of `Analyses` (the plural form of `Analysis`)

File: python/revng/api/target.py [hunk 25]
```diff
@@ -2,38 +2,38 @@
...
 
 class Target:
-    def __init__(self, target, container: Container):
+    def __init__(self, target, container):
         self._target = target
         self._container = container
 
     @staticmethod
-    def create(kind: Kind, container: Container, path_components: List[str]) -> Optional["Target"]:
+    def create(_kind, container: str, path_components: List[str]) -> Optional["Target"]:
```
[the remark this answers is not in the document]
Comment (mrjackv) [reply python/revng/api/target.py:19]: As an undocumented convention, variables that start with `_` are used for pointer (e.g. the `_kind` here refers to an `rp_kind *`, not a string or an object)

File: python/revng/daemon/graphql.py (hunk 19 above)
Comment (mrjackv) [reply python/revng/daemon/graphql.py:1]: It was renamed from `static_handlers.py`, but since a lot of code has been dropped from it Github hasn't detected it as a rename

File: include/revng/PipelineDescriptor/AnalysesList.h (hunk 24 above)
[the remark this answers is not in the document]
Comment (aleclearmind) [reply include/revng/PipelineDescriptor/AnalysesList.h:8]: In English it's preferable to use "cat list" instead of "cats list" ([random source](https://www.proz.com/forum/linguistics/245800-noun_list_is_noun_singular_or_plural.html)) and that's the practice we adopt elsewhere too.

Anyway, if we already have other old code using plural, it makes sense to be coherent. But in the future let's try to follow this practice.

File: include/revng/PipelineDescriptor/Kind.h (hunk 21 above)
[the remark this answers is not in the document]
Comment (aleclearmind) [reply include/revng/PipelineDescriptor/Kind.h:18]: Can we rename `using Parent` to `using Base` or `using BaseClass`?

File: python/revng/api/target.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (aleclearmind) [reply python/revng/api/target.py:19]: Mmmh, OK, but I think in Python `_` is used to hint that a field is """private""", isn't it? Not sure it's the best convention.

Anyway, we can consider reverting the convention in another moment.

File: include/revng/PipelineDescriptor/Kind.h (hunk 21 above)
[the remark this answers is not in the document]
Comment (mrjackv) [reply include/revng/PipelineDescriptor/Kind.h:18]: Sure, I was trying to be the least invasive when integrating pipelineDescriptor

File: include/revng/PipelineDescriptor/Analysis.h (hunk 2 above)
Comment (mrjackv) [reply include/revng/PipelineDescriptor/Analysis.h:16]: 1. The vector contains `AnalysisArgument`, which contains the name and type
2. See (1)

Comment (aleclearmind) [reply include/revng/PipelineDescriptor/Analysis.h:16]: https://github.com/revng/revng/pull/318#discussion_r1302856782

File: include/revng/PipelineDescriptor/AnalysisArgument.h (hunk 3 above)
Comment (mrjackv) [reply include/revng/PipelineDescriptor/AnalysisArgument.h:8]: > Backport it

Will do

> Also, these are the accepted kinds, they don't have anything to do with analysis arguments.

There are 2 different things:
* An analysis argument, which needs to point to a target in the pipeline to produce a result (e.g. `ImportBinary` has an argument `input` which accepts kind `Binary`)
* An analysis option, which (AFAIK) did not have any API in PipelineC to introspect them since (at the time of writing the API) no analysis had them

File: include/revng/Pipeline/Step.h [hunk 26]
```diff
@@ -145,6 +145,13 @@ class Step {
     }
   }
 
+  const std::string getArtifactContainerName() const {
```
[the remark this answers is not in the document]
Comment (mrjackv) [reply include/revng/Pipeline/Step.h]: Done (1) and (2)

File: include/revng/PipelineDescriptor/AnalysesList.h [hunk 27]
```diff
@@ -0,0 +1,32 @@
...
+doc: Data structure representing a list of analyses
+type: struct
+fields:
+  - name: Name
+    doc: AnalysesList' name
```
[the remark this answers is not in the document]
Comment (mrjackv) [reply include/revng/PipelineDescriptor/AnalysesList.h]: Fixed

File: include/revng/PipelineDescriptor/PipelineDescriptor.h (hunk 1 above)
Comment (mrjackv) [reply include/revng/PipelineDescriptor/PipelineDescriptor.h]: Fixed

File: include/revng/PipelineDescriptor/Container.h [hunk 28]
```diff
@@ -0,0 +1,30 @@
...
+
+/* TUPLE-TREE-YAML
+name: Container
+doc: Data structure representing a Container
+type: struct
+fields:
+  - name: Name
+    doc: Container's name
+    type: string
+  - name: Mime
```
[the remark this answers is not in the document]
Comment (mrjackv) [reply include/revng/PipelineDescriptor/Container.h]: Renamed

File: include/revng/PipelineDescriptor/Kind.h (hunk 21 above)
[the remark this answers is not in the document]
Comment (mrjackv) [reply include/revng/PipelineDescriptor/Kind.h]: Renamed to `BaseClass`, changed KindParent, StepParent and RankParent to just Parent

File: include/revng/Pipes/DescriptorConverter.h [hunk 29]
```diff
@@ -0,0 +1,11 @@
...
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "revng/PipelineDescriptor/PipelineDescriptor.h"
+#include "revng/Pipes/PipelineManager.h"
+
+pipelineDescriptor::PipelineDescriptor
+generateDescriptor(const revng::pipes::PipelineManager &Manager);
```
[the remark this answers is not in the document]
Comment (mrjackv) [reply include/revng/Pipes/DescriptorConverter.h]: Moved to a method of Runner (`Runner::description()`)

File: include/revng/Pipes/PipelineManager.h (hunk 5 above)
Comment (mrjackv) [reply include/revng/Pipes/PipelineManager.h]: Dropped

File: lib/Pipes/DescriptorConverter.cpp (hunk 6 above)
Comment (mrjackv) [reply lib/Pipes/DescriptorConverter.cpp]: Renamed

File: scripts/tuple_tree_generator/tuple_tree_generator/templates/struct_impl.cpp.tpl (hunk 11 above)
Comment (mrjackv) [reply scripts/tuple_tree_generator/tuple_tree_generator/templates/struct_impl.cpp.tpl:11]: Done

File: python/revng/tupletree.py (hunk 10 above)
Comment (mrjackv) [reply python/revng/tupletree.py:331]: Split off

File: include/revng/Pipeline/ContainerFactorySet.h (hunk 14 above)
Comment (mrjackv) [reply include/revng/Pipeline/ContainerFactorySet.h]: Renamed to `at`

File: include/revng/PipelineDescriptor/Analysis.h (hunk 16 above)
Comment (mrjackv) [reply include/revng/PipelineDescriptor/Analysis.h]: Fixed

File: lib/Pipes/DescriptorConverter.cpp (hunk 17 above)
Comment (mrjackv) [reply lib/Pipes/DescriptorConverter.cpp]: Fixed

File: python/revng/pipelineDescriptor/__init__.py (hunk 9 above)
Comment (mrjackv) [reply python/revng/pipelineDescriptor/__init__.py]: Changed

File: typescript/CMakeLists.txt (hunk 13 above)
Comment (mrjackv) [reply typescript/CMakeLists.txt]: Changed

File: lib/Pipes/PipelineManager.cpp (hunk 7 above)
Comment (mrjackv) [reply lib/Pipes/PipelineManager.cpp]: Done

File: include/revng/PipelineDescriptor/PipelineDescriptor.h (hunk 4 above)
Comment (mrjackv) [reply include/revng/PipelineDescriptor/PipelineDescriptor.h]: 1. Done
2. Renamed to `Description`

File: include/revng/PipelineDescriptor/Kind.h [hunk 30]
```diff
@@ -0,0 +1,43 @@
...
+  - name: PreferredKinds
+    doc: The Kinds that this Kind prefers
```
[the remark this answers is not in the document]
Comment (mrjackv) [reply include/revng/PipelineDescriptor/Kind.h]: Improved doc string

File: include/revng/PipelineC/Prototypes.h (hunk 15 above)
Comment (mrjackv) [reply include/revng/PipelineC/Prototypes.h]: Fixed

File: python/revng/api/target.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (mrjackv) [reply python/revng/api/target.py:19]: > Mmmh, OK, but I think in Python _ is used to hint that a field is """private""", isn't it?

That's (mostly) for classes, also it's more like "please don't touch this unless you're sure". Double underscore invokes mangling which is usually when you want to avoid people touching the attribute

> Not sure it's the best convention.

From PEP8:

> _single_leading_underscore: weak “internal use” indicator

It's a bit of a stretch, but pointers are basically internal to `revng.api`, so it "makes sense" (IMHO)

File: include/revng/PipelineDescriptor/AnalysisArgument.h (hunk 3 above)
Comment (mrjackv) [reply include/revng/PipelineDescriptor/AnalysisArgument.h]: Changed the name to `AnalysisContainerInput`, also added Options

File: lib/Pipes/PipelineManager.cpp (hunk 8 above)
Comment (mrjackv) [reply lib/Pipes/PipelineManager.cpp]: 1. Ok
2. Moved some code around, there's no more duplication

File: include/revng/Pipeline/Description/PipelineDescription.h [hunk 31]
```diff
@@ -0,0 +1,50 @@
...
+      elementType: Container
+  - name: Steps
+    sequence:
+      type: SortedVector
+      elementType: Step
+TUPLE-TREE-YAML */
+
+#include "revng/Pipeline/Description/Generated/Early/PipelineDescription.h"
+
+class pipelineDescription::PipelineDescription
```
Comment (aleclearmind) [inline include/revng/Pipeline/Description/PipelineDescription.h:44]: This is should be part of the `pipeline` namespace (which is not `revng::pipeline`).
Shall we go for `pipeline::description::{Pipeline,Container,Rank,...}`?

**Comment by aleclearmind** [comment]
Merged in f95ef2b. Thanks.
